modules/substrate/src/verifier.rs

```diff
@@ -0,0 +1,667 @@
+use sp_runtime::generic::OpaqueDigestItemId;
+use sp_runtime::traits::{CheckedAdd, Header as HeaderT};
+use sp_std::{prelude::Vec, vec};
+
+/// The finality proof used by the pallet.
+///
+/// For a Substrate based chain using Grandpa this will
+/// be an encoded Grandpa Justification.
+pub type FinalityProof<'a> = &'a [u8];
+
+/// Errors which can happen while importing a header.
+#[derive(Debug, PartialEq)]
+pub enum ImportError {
+	/// This header is older than our latest finalized block, thus not useful.
+	OldHeader,
+	/// This header has already been imported by the pallet.
+	HeaderAlreadyExists,
+	/// This header has never been imported by the pallet.
+	UnknownHeader,
+	/// We're missing a parent for this header.
+	MissingParent,
+	/// We were unable to prove finality for this header.
+	UnfinalizedHeader,
+	/// We failed to verify this header's ancestry.
+	AncestryCheckFailed,
+	/// The header is missing digests related to consensus events.
+	///
+	/// This will typically have to do with missing authority set change signals.
+	MissingConsensusDigest,
+	/// Trying to prematurely import a justification
+	PrematureJustification,
+	/// The height of the next authority set change overflowed.
+	ScheduledHeightOverflow,
+}
+
+/// A trait for verifying whether a header is valid for a particular blockchain.
+pub trait ChainVerifier<S, H: HeaderT> {
+	/// Import a header to the pallet.
+	fn import_header(storage: &mut S, header: H) -> Result<(), ImportError>;
+
+	/// Verify that the given header has been finalized and is part of the canonical chain.
+	fn verify_finality(storage: &mut S, hash: H::Hash, proof: FinalityProof) -> Result<(), ImportError>;
+}
+
+/// Used to verify imported headers and their finality status.
+#[derive(Debug)]
+pub struct Verifier;
+
+impl<S, H> ChainVerifier<S, H> for Verifier
+where
+	S: BridgeStorage<Header = H>,
+	H: HeaderT,
+{
+	fn import_header(storage: &mut S, header: H) -> Result<(), ImportError> {
+		let best_finalized = storage.best_finalized_header();
+
+		if header.number() < best_finalized.number() {
+			return Err(ImportError::OldHeader);
+		}
+
+		if storage.header_exists(header.hash()) {
+			return Err(ImportError::HeaderAlreadyExists);
+		}
+
+		let parent_header = storage.get_header_by_hash(*header.parent_hash());
+		if parent_header.is_none() {
+			return Err(ImportError::MissingParent);
+		}
+
+		// We don't need the justification right away, but we should note it.
+		let requires_justification = find_scheduled_change(&header).is_some();
+		let is_finalized = false;
+		storage.write_header(&ImportedHeader::new(header, requires_justification, is_finalized));
+
+		Ok(())
+	}
+
+	fn verify_finality(storage: &mut S, hash: H::Hash, proof: FinalityProof) -> Result<(), ImportError> {
+		// Make sure that we've previously imported this header
+		let header = storage.get_header_by_hash(hash).ok_or(ImportError::UnknownHeader)?;
+
+		let current_authority_set = storage.current_authority_set();
```

Comment:
It seems weird we don't verify that the `header` actually should be finalized by `current_authority_set`. i.e. It seems this code assumes that `next_validator_set_enactment_block_number > header.number() > last_finalized_header_number`, but I don't think it's clear where it's enforced (and why not here).

Reply:
> It seems this code assumes that next_validator_set_enactment_block_number > header.number() > last_finalized_header_number, but I don't think it's clear where it's enforced (and why not here).

I think there is some work left to be done around the authority set changes (and its assumptions), but I'm going to address that in a follow up PR if that's alright.

> It seems weird we don't verify that the header actually should be finalized by current_authority_set

I was under the assumption that this would at least partly be handled by the function which checks the finality proofs but maybe that's not the correct thing to do. I'll need to revisit that once I start integrating it into the pallet.